pkg/datastore/filedb/filedb.go

```diff
@@ -61,11 +65,100 @@ func (f *FileDB) fetch(ctx context.Context, path string) ([]byte, error) {
 }
 
 func (f *FileDB) Find(ctx context.Context, col datastore.Collection, opts datastore.ListOptions) (datastore.Iterator, error) {
-	_, ok := col.(datastore.ShardStorable)
+	scol, ok := col.(datastore.ShardStorable)
 	if !ok {
 		return nil, datastore.ErrUnsupported
 	}
-	return nil, datastore.ErrUnimplemented
+
+	var (
+		kind   = col.Kind()
+		shards = scol.ListInUsedShards()
+		// Map of objects values with the first key is the object id.
+		objects map[string][][]byte
+	)
+
+	// Fetch the first part of all objects under a specified "directory".
+	for _, shard := range shards {
+		dpath := makeHotStorageDirPath(kind, shard)
+		parts, err := f.backend.List(ctx, dpath)
+		if err != nil {
+			f.logger.Error("failed to find entities",
+				zap.String("kind", kind),
+				zap.Error(err),
+			)
+			return nil, err
+		}
+
+		if objects == nil {
+			objects = make(map[string][][]byte, len(parts))
+		}
+		for _, obj := range parts {
+			id := filepath.Base(obj.Path)
+
+			// Check if the raw data stored under this path is fetched or not
+			// based on its etag value stored as key in `f.cache` store.
+			cdata, err := f.cache.Get(obj.Etag)
+			if err == nil {
+				objects[id] = append(objects[id], cdata.([]byte))
+				continue
+			}
```

Comment:
How about grouping and storing all models of a kind using HashCache https://github.com/pipe-cd/pipecd/blob/master/pkg/cache/rediscache/hashcache.go?
Then we can fetchAll to check instead of calling every time like this.

Reply:
And I think we need to design the key name carefully instead of directly using the `Etag` value to avoid conflict with other keys in Redis. At other places we're prefixing the name.

Reply:
I feel you, in that case, it's possible to have some objects updated while the whole others are not, meaning that we still for loop to check which one is updated and which one is not. In that case, we have a minor point to care about: Since I store the whole object content as the value for etag key, if we group it as a single value on use GetAll, it's a bit dangerous in case there are a lot of objects. Of course, as the downside of the current implementation, we may have a bunch of requests to cache to check whether the object is updated or not based on its etag. Wdyt about his trade-off 🤔

Reply:
> And I think we need to design the key name carefully instead of directly using the Etag value to avoid conflict with other keys in Redis. At other places we're prefixing the name.

Nice catch, lets me address it 🙆‍♂️

Reply:
> it's possible to have some objects updated while the whole others are not

Yes, it is. For the outdated ones, we call Cache HSET to update after fetching directly from the file store.

> if we group it as a single value on use GetAll, it's a bit dangerous in case there are a lot of objects

I see. Storing by HashCache, I mean the Field Key is the object ID instead of eTag, eTag is included in the value with object content.  So the number of objects will not increase when the object is updated.
When storing by ETag the number of objects in cache will increase quickly whenever an object has been updated. In that case, we will need TTL or something like that to deal with that problem.

Wdyt?

Reply:
Lets me confirm one thing: using hashcache to store mean we will use object `id_shard` as hashkey and the value will be kind of `{etag: etag_value, data: data}`: in this case, we have the number of HashCache equal to the `number of object * shard`. Or we will only have one hashkey like kind and the value will be `{ object_id_1: {etag: etag_value, data: data}}`?

if it's the first case, I can't see the point of using HashCache, normal cache with the key is the `objectId_shard` is good as well.

Reply:
> When storing by ETag the number of objects in cache will increase quickly whenever an object has been updated. In that case, we will need TTL or something like that to deal with that problem.

You have the point, I think the same about that, and tunning the TTL value so that the cache works well is not easy.  Better to use `update` instead of `adding` new value with key is the `etag` as you said 👍

Reply:
> Lets me confirm one thing: using hashcache to store mean we will use object id_shard as hashkey and the value will be kind of {etag: etag_value, data: data}: in this case, we have the number of HashCache equal to the number of object * shard. Or we will only have one hashkey like kind and the value will be { object_id_1: {etag: etag_value, data: data}}?

I don't think so. For using hash cache we need 2 kinds of key: Hash Key and Field Key.
Hash Key can be `FILEDB_SHARD_{ENTITY_KIND}_{SHARD_NAME}_{ENTITY_PROJECT}`
Field Key can be `{ENTITY_ID}`
Field Value can be `{etag: etag_value, data: data}`.

Reply:
> Or we will only have one hashkey like kind and the value will be { object_id_1: {etag: etag_value, data: data}}?

Yes, it's my second suggestion, we're on the same page. But there is a problem with this approach, that's the size of the value on this hashkey can be too large, since the number of objects can be big and we store the whole object's content as `FieldValue.data`

Reply:
How about this
- Normal cache (not hashcache)
- FieldKey: `entity_id_shard`
- FieldValue: `{etag: etag_value, data: data}`
🤔

Reply:
What I am concerned about that way is that it still requires N requests to our Redis.
But definitely, it is simpler so let's apply that way for now and see how it goes. 👍

Reply:
> What I am concerned about that way is that it still requires N requests to our Redis.

I feel you, then why not both 🤔 
I mean we can make our cache a bit more complicated to handle this case, we will have:
hashcache as your drafted with:
- hashkey: List_{kind}
- field key: entityId_shard
- field value: etagValue
cache to store object data
- key: etag_entityId (I mean `etag_` is the prefix not the value of etag)
- value: {etag: etagValue, data: data}

And whenever we update the etagValue in the hashcache, we update the cache storing object content as well. Wdyt 👀

Reply:
oops, I forgot that we still need to fetch cache to get the content by separated cache request. Please forget the above suggestion 🙏

Reply:
Updated, PTAL when you have time 😉